lib/fluxsink/a2rfluxsink.cc

```diff
@@ -0,0 +1,218 @@
+#include "globals.h"
+#include "flags.h"
+#include "fluxmap.h"
+#include "bytes.h"
+#include "protocol.h"
+#include "fluxsink/fluxsink.h"
+#include "decoders/fluxmapreader.h"
+#include "lib/fluxsink/fluxsink.pb.h"
+#include "lib/logger.h"
+#include "proto.h"
+#include "fmt/format.h"
+#include "fmt/chrono.h"
+#include "fluxmap.h"
+#include "a2r.h"
+#include <fstream>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+namespace
+{
+uint32_t ticks_to_a2r(uint32_t ticks) {
+    return ticks * NS_PER_TICK / A2R_NS_PER_TICK;
+}
+
+bool singlesided(void) {
+	return config.heads().start() == config.heads().end();
+}
+
+class A2RFluxSink : public FluxSink
+{
+public:
+	A2RFluxSink(const A2RFluxSinkProto& lconfig):
+		_config(lconfig),
+		_bytes{},
+		_writer{_bytes.writer()}
+	{
+
+		Logger() << fmt::format("A2R: writing A2R {} file containing {} tracks\n",
+			singlesided() ? "single sided" : "double sided",
+			config.tracks().end() - config.tracks().start() + 1
+		);
+
+		time_t now{std::time(nullptr)};
+		auto t = gmtime(&now);
+		_metadata["image_date"] = fmt::format("{:%FT%TZ}", *t);
+	}
+
+	~A2RFluxSink()
+	{
+		writeHeader();
+		writeInfo();
+		writeStream();
+		writeMeta();
+
+		Logger() << "A2R: writing output file...\n";
+		std::ofstream of(_config.filename(), std::ios::out | std::ios::binary);
+		if (!of.is_open())
+			Error() << "cannot open output file";
+		_bytes.writeTo(of);
+		of.close();
+	}
+
+private:
+	void writeData(const auto data, size_t size) {
+		_writer += Bytes(reinterpret_cast<const uint8_t*>(data), size);
+	}
+	void writeChunkAndData(uint32_t chunk_id, const auto data, size_t size) {
+		_writer.write_le32(chunk_id);
+		_writer.write_le32(size);
+		writeData(data, size);
```

Comment:
You can't write structures containing non-single-byte values like this. (The order in disk will vary depending on whether you're on a LE/BE system.)